corehq/motech/openmrs/workflow.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import absolute_import
+
+from collections import namedtuple
+
+
+class WorkflowTask(object):
```

Comment:
I still think it's much cleaner to create subclasses rather than pass in functions and args. In the current implementation the forward and rollback functions are separated and can't share any data easily. They also often require some of the same args so you end up passing them in twice.

You also end up with the weird logic of passing a task result to the rollback task via the kwargs. If both were contained within the same class you could just set a class property as the result which the rollback method could access. You also wouldn't need the complicated `returns_result` and `returns_subtasks` logic.

Finally it would make it much easier to read:
```
WorkflowTask(
    func=create_encounter,
    func_args=(requests, person_uuid, provider_uuid, start_datetime, values_for_concept,
        encounter_type, openmrs_form, visit_uuid, location_uuid),
    rollback_func=delete_encounter,
    rollback_args=(requests, ),
    returns_result=True,
    returns_subtasks=True
)
```
vs
```
CreateEncounterTask(
    requests, person_uuid, provider_uuid, start_datetime, values_for_concept,
    encounter_type, openmrs_form, visit_uuid, location_uuid
)
```

Reply:
Thanks. Those are good points, and that is a lot more readable.